### ISSUE TITLE
Incorrect inferred type for filter(None, Iterator)

### ISSUE THREAD

**Issue by LiraNuna** [python/mypy#3835]
mypy incorrectly infers the type of the iterator returned by `filter` when the first param is always `None` as still `Optional`:

```python
reveal_type(filter(None, ['test', None, 'bla']))  # Revealed type is 'typing.Iterator[Union[builtins.str, builtins.None]]'
reveal_type(filter(lambda x: x is not None, ['test', None, 'bla']))  # Revealed type is 'typing.Iterator[Union[builtins.str, builtins.None]]'
```

This only happens when `--strict-optional` flag is passed.
I'm not sure if this is a bug or an impossibility of inferring the semantics of `filter`, since in theory it only reduces the count of the iterator and does not necessarily modifies the content of it.

However I could make the argument that a special case could be added to `filter(None, ...)` since the docs specifically mention this behavior:

>  If _function_ is `None`, the identity function is assumed, that is, all elements of iterable that are false are removed.

**Comment by OddBloke** [comment]
Reversing the order of the overload definitions in typeshed does solve this problem:

```py
@overload
def filter(function: None, iterable: Iterable[Optional[_T]]) -> Iterator[_T]: ...
@overload
 def filter(function: Optional[Callable[[_T], Any]],
            iterable: Iterable[_T]) -> Iterator[_T]: ...
```

but it looks like this has been done before in https://github.com/python/typeshed/commit/beb9183103e6cdc5805e64e0f88d7ab80b031cee  (admittedly before Optional was added to the non-None definition), so I don't know if there's something more complex going on here.

**Comment by LiraNuna** [comment]
This has regressed in 0.520. I tested and it works correctly in 0.511

**Comment by LiraNuna** [comment]
This is still an issue in 0.540

**Comment by LiraNuna** [comment]
This seems fixed in 0.600. Closing!
